File: CS/Task26/RepositoryItemMyDateEdit.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Controls;
using DevExpress.Utils;

namespace Task26
{
   [UserRepositoryItem("RegisterMyDateEdit")]
    public class RepositoryItemMyDateEdit: RepositoryItemDateEdit
    {
        List<Note> fNotes;
        List<Note> fActiveNotes;
        NoteForm fNoteForm;
        SelectForm fSelectForm;
        
       static RepositoryItemMyDateEdit() { RegisterMyDateEdit(); }

        public RepositoryItemMyDateEdit()
        {
            fNotes = new List<Note>();
            fActiveNotes = new List<Note>();
            fSelectForm = new SelectForm();
        }

        public const string MyDateEditName = "MyDateEdit";

       public bool IsVistaDisplay
       {
           get { return IsVistaDisplayMode(); }
       }
       
       public override string EditorTypeName { get { return MyDateEditName; } }
        
        public static void RegisterMyDateEdit()
        {      
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MyDateEditName, 
              typeof(MyDateEdit),
              typeof(RepositoryItemMyDateEdit),
              typeof(MyDateEditViewInfo),
              new MyDateEditPainter(),
              true));
        }

        public virtual List<Note> Notes
        {
            get { return fNotes; }
        }

        public virtual List<Note> ActiveNotes
        {
            get { return fActiveNotes; }
        }

        protected virtual DateTime EditorDate
        {
            get
            {
                if (OwnerEdit == null)
                    return new DateTime();
                else
                    return OwnerEdit.DateTime;
            }
        }

        public bool HaveActiveNotes
        {
            get { return ActiveNotes.Count > 0; }
        }

        public virtual List<Note> GetDateNotes(DateTime Date)
        {
            List<Note> Res = new List<Note>();

            foreach (Note N in Notes)
            {
                if (N.Date.Date == Date.Date)
                {
                    Res.Add(N);
                }
            }

            return Res;
        }

       public void AddNote(Note N)
       {
           Notes.Add(N);
           CheckNotes();
       }
       
       public void AddNote()
        {
            AddNote(EditorDate);
        }
       
       public virtual void AddNote(DateTime Date)
       {
           Note NewNote = new Note(Date.Date);
           fNoteForm = new NoteForm(NewNote);
           if (fNoteForm.ShowDialog() == DialogResult.OK)
           {
               Notes.Add(NewNote);
               CheckNotes();
           }
           fNoteForm.Dispose();
       }

        public void EditNote(Note N)
        {
            int Index = Notes.IndexOf(N);
            if (Index >= 0)
                EditNote(Index);
        }
       
       public virtual void EditNote(int Index)
       {
           if ((Index >= 0) && (Index < Notes.Count))
           {
               Note Tmp = Notes[Index];
               fNoteForm = new NoteForm(Tmp);
               if (fNoteForm.ShowDialog() == DialogResult.OK)
               {
                   Notes[Index] = Tmp;
                   CheckNotes();
               }
               fNoteForm.Dispose();
           }
       }

        public virtual void EditNotes(DateTime Date)
        {
            List<Note> LN = GetDateNotes(Date);

            if (LN.Count > 0)
            {
                if (LN.Count == 1)
                {
                    EditNote(LN[0]);
                }
                else
                {
                    int SelIndex = fSelectForm.SelectNote(LN);
                    EditNote(SelIndex);
                }
            }
        }

        public virtual bool HaveNote(DateTime Date)
        {
            foreach (Note N in Notes)
            {
                if (N.Date.Date == Date.Date)
                {
                    return true;
                }
            }
            return false;
        }
       
       public virtual void DeleteNotes(DateTime Date)
        {
            List<Note> LN = GetDateNotes(Date);

            if (LN.Count > 0)
            {
                if (LN.Count == 1)
                {
                    Notes.Remove(LN[0]);
                }
                else
                {
                    List<Note> Selection = fSelectForm.SelectNotes(LN);
                    foreach (Note N in Selection)
                    {
                        if (ActiveNotes.Contains(N))
                            ActiveNotes.Remove(N);
                        Notes.Remove(N);
                    }
                }
                CheckNotes();
            }
        }

        public virtual void CheckNotes()
        {
            List<Note> LN = GetDateNotes(EditorDate);

            for (int i = ActiveNotes.Count-1; i >= 0; i--)
            {
                if (ActiveNotes[i].Date.Date != EditorDate.Date)
                    ActiveNotes.RemoveAt(i);
            }

            foreach (Note N in LN)
            {
                if ((N.Date.Date == EditorDate.Date) && !ActiveNotes.Contains(N))
                {
                    ActiveNotes.Add(N);
                }
            }
        }

       protected internal virtual void OnNoteIconClicked()
       {
           if (ActiveNotes.Count > 0)
           {
               int Index = (ActiveNotes.Count == 1) ? 0 : fSelectForm.SelectNote(ActiveNotes);
               EditNote(ActiveNotes[Index]);
           }
       }

        public override BaseEditViewInfo CreateViewInfo()
        {
            return new MyDateEditViewInfo(this);
        }
       
       public override void Assign(RepositoryItem item)
        {
            BeginUpdate(); 
            try
            {
                base.Assign(item);
                RepositoryItemMyDateEdit source = item as RepositoryItemMyDateEdit;
                if(source == null) return;

                fNotes = source.fNotes;
            }
            finally
            {
                EndUpdate();
            }
        }

       protected override void  Dispose(bool disposing)
       {
 	       base.Dispose(disposing);
           if (disposing)
           {
               fSelectForm.Dispose();
           }
       }
    }
}

File: VB/Task26/RepositoryItemMyDateEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Utils

Namespace Task26
   <UserRepositoryItem("RegisterMyDateEdit")> _
   Public Class RepositoryItemMyDateEdit
	   Inherits RepositoryItemDateEdit
		Private fNotes As List(Of Note)
		Private fActiveNotes As List(Of Note)
		Private fNoteForm As NoteForm
		Private fSelectForm As SelectForm

	   Shared Sub New()
		   RegisterMyDateEdit()
	   End Sub

		Public Sub New()
			fNotes = New List(Of Note)()
			fActiveNotes = New List(Of Note)()
			fSelectForm = New SelectForm()
		End Sub

		Public Const MyDateEditName As String = "MyDateEdit"

	   Public ReadOnly Property IsVistaDisplay() As Boolean
		   Get
			   Return IsVistaDisplayMode()
		   End Get
	   End Property

	   Public Overrides ReadOnly Property EditorTypeName() As String
		   Get
			   Return MyDateEditName
		   End Get
	   End Property

		Public Shared Sub RegisterMyDateEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MyDateEditName, GetType(MyDateEdit), GetType(RepositoryItemMyDateEdit), GetType(MyDateEditViewInfo), New MyDateEditPainter(), True))
		End Sub

		Public Overridable ReadOnly Property Notes() As List(Of Note)
			Get
				Return fNotes
			End Get
		End Property

		Public Overridable ReadOnly Property ActiveNotes() As List(Of Note)
			Get
				Return fActiveNotes
			End Get
		End Property

		Protected Overridable ReadOnly Property EditorDate() As DateTime
			Get
				If OwnerEdit Is Nothing Then
					Return New DateTime()
				Else
					Return OwnerEdit.DateTime
				End If
			End Get
		End Property

		Public ReadOnly Property HaveActiveNotes() As Boolean
			Get
				Return ActiveNotes.Count > 0
			End Get
		End Property

		Public Overridable Function GetDateNotes(ByVal [Date] As DateTime) As List(Of Note)
			Dim Res As New List(Of Note)()

			For Each N As Note In Notes
				If N.Date.Date = [Date].Date Then
					Res.Add(N)
				End If
			Next N

			Return Res
		End Function

	   Public Sub AddNote(ByVal N As Note)
		   Notes.Add(N)
		   CheckNotes()
	   End Sub

	   Public Sub AddNote()
			AddNote(EditorDate)
	   End Sub

	   Public Overridable Sub AddNote(ByVal [Date] As DateTime)
		   Dim NewNote As New Note([Date].Date)
		   fNoteForm = New NoteForm(NewNote)
		   If fNoteForm.ShowDialog() = DialogResult.OK Then
			   Notes.Add(NewNote)
			   CheckNotes()
		   End If
		   fNoteForm.Dispose()
	   End Sub

		Public Sub EditNote(ByVal N As Note)
			Dim Index As Integer = Notes.IndexOf(N)
			If Index >= 0 Then
				EditNote(Index)
			End If
		End Sub

	   Public Overridable Sub EditNote(ByVal Index As Integer)
		   If (Index >= 0) AndAlso (Index < Notes.Count) Then
			   Dim Tmp As Note = Notes(Index)
			   fNoteForm = New NoteForm(Tmp)
			   If fNoteForm.ShowDialog() = DialogResult.OK Then
				   Notes(Index) = Tmp
				   CheckNotes()
			   End If
			   fNoteForm.Dispose()
		   End If
	   End Sub

		Public Overridable Sub EditNotes(ByVal [Date] As DateTime)
			Dim LN As List(Of Note) = GetDateNotes([Date])

			If LN.Count > 0 Then
				If LN.Count = 1 Then
					EditNote(LN(0))
				Else
					Dim SelIndex As Integer = fSelectForm.SelectNote(LN)
					EditNote(SelIndex)
				End If
			End If
		End Sub

		Public Overridable Function HaveNote(ByVal [Date] As DateTime) As Boolean
			For Each N As Note In Notes
				If N.Date.Date = [Date].Date Then
					Return True
				End If
			Next N
			Return False
		End Function

	   Public Overridable Sub DeleteNotes(ByVal [Date] As DateTime)
			Dim LN As List(Of Note) = GetDateNotes([Date])

			If LN.Count > 0 Then
				If LN.Count = 1 Then
					Notes.Remove(LN(0))
				Else
					Dim Selection As List(Of Note) = fSelectForm.SelectNotes(LN)
					For Each N As Note In Selection
						If ActiveNotes.Contains(N) Then
							ActiveNotes.Remove(N)
						End If
						Notes.Remove(N)
					Next N
				End If
				CheckNotes()
			End If
	   End Sub

		Public Overridable Sub CheckNotes()
			Dim LN As List(Of Note) = GetDateNotes(EditorDate)

			For i As Integer = ActiveNotes.Count-1 To 0 Step -1
				If ActiveNotes(i).Date.Date <> EditorDate.Date Then
					ActiveNotes.RemoveAt(i)
				End If
			Next i

			For Each N As Note In LN
				If (N.Date.Date = EditorDate.Date) AndAlso (Not ActiveNotes.Contains(N)) Then
					ActiveNotes.Add(N)
				End If
			Next N
		End Sub

	   Protected Friend Overridable Sub OnNoteIconClicked()
		   If ActiveNotes.Count > 0 Then
			   Dim Index As Integer = If((ActiveNotes.Count = 1), 0, fSelectForm.SelectNote(ActiveNotes))
			   EditNote(ActiveNotes(Index))
		   End If
	   End Sub

		Public Overrides Function CreateViewInfo() As BaseEditViewInfo
			Return New MyDateEditViewInfo(Me)
		End Function

	   Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemMyDateEdit = TryCast(item, RepositoryItemMyDateEdit)
				If source Is Nothing Then
					Return
				End If

				fNotes = source.fNotes
			Finally
				EndUpdate()
			End Try
	   End Sub

	   Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			MyBase.Dispose(disposing)
		   If disposing Then
			   fSelectForm.Dispose()
		   End If
	   End Sub
   End Class
End Namespace
